### PR TITLE
enables copying rds snapshots across regions

### PR DESCRIPTION
**Pull request by brockhaywood**
Related module: https://github.com/ansible/ansible-modules-extras/pull/1281 which enables searching for rds snapshots

### CODE REVIEW

**Comment by robynbergeron** [comment]
Hi @brockhaywood --

Thanks for submitting this new module to Ansible Extras! This module is now in community review, a process that is open to all Ansible users. In order for this module to be approved, it must gain the following votes, which can be contributed by anyone other than the author of the new module:

“works_for_me”: If you have tested the module thoroughly, including testing of all of the module’s options, and if the module works for you, please add “works_for_me” in the comments.

“passes_guidelines”: If you have gone through the module guidelines and the module meets all of the requirements, please add “passes_guidelines” in the comments. Guidelines are available here: http://docs.ansible.com/developing_modules.html#module-checklist

“needs_revision”: If the module fails to work for you, or if it doesn’t meet guidelines, please add “needs_revision” in the comments with details about what needs to be fixed.

When a module has both “works_for_me” and “passes_guidelines” tags, we will promote the module for inclusion in Ansible Extras. At this point, you will be expected to maintain the module by fixing bugs and evaluating pull requests in a timely manner.

More information about the review process can be seen here: 
https://github.com/ansible/ansible-modules-extras/blob/devel/REVIEWERS.md

Thanks again for submitting your Ansible module!

**Comment by robynbergeron** [comment]
Pinging AWS friends for feedback: @jsmartin @scicoin-project @tombamford @garethr @jarv @jsdalton @silviud @adq @zbal @zeekin @willthames @lwade @carsongee @defionscode
@tastychutney @bpennypacker @loia

Also note there is another related new module for RDS, #1281

File: cloud/amazon/rds_snapshot_copy.py [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+    # Retrieve any AWS settings from the environment.
+    region, ec2_url, aws_connect_kwargs = get_aws_connection_info(module)
+
+    if not region:
+        module.fail_json(msg = str("Either region or AWS_REGION or EC2_REGION "
+                                   "environment variable or boto config "
+                                   "aws_region or ec2_region must be set."))
+
+    try:
+        conn = boto.rds.connect_to_region(region, **aws_connect_kwargs)
```
Comment (willthames) [inline cloud/amazon/rds_snapshot_copy.py]: please use `connect_to_aws(boto.rds, region, **aws_connect_kwargs)` 

(see ansible/ansible-modules-core#2421 for details)

File: cloud/amazon/rds_snapshot_copy.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+
+        if source_region and source_region != region:
+            source_conn = boto.rds.connect_to_region(
```
Comment (willthames) [inline cloud/amazon/rds_snapshot_copy.py]: again, use `connect_to_aws`

File: cloud/amazon/rds_snapshot_copy.py [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+except ImportError:
+    HAS_BOTO = False
+
+
+def copy_snapshot(module, source_conn, conn,
+                  source_db_snapshot_identifier=None,
+                  target_db_snapshot_identifier=None):
+
+    if not source_db_snapshot_identifier or not target_db_snapshot_identifier:
+        module.fail_json(msg='Source DB Snapshot Id and Target DB Snapshot '
```
Comment (willthames) [inline cloud/amazon/rds_snapshot_copy.py:72]: enforce this with `required=True` in the `argument_spec`

File: cloud/amazon/rds_snapshot_copy.py [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+            source_db_snapshot_identifier = dict(),
+            target_db_snapshot_identifier = dict(),
+        )
+    )
+    module = AnsibleModule(argument_spec=argument_spec)
+    return module
+
+
+def main():
+    module = copy_rds_snapshot_ansible_module()
```
Comment (willthames) [inline cloud/amazon/rds_snapshot_copy.py]: I don't see the point of this abstraction - few other modules do this.

File: cloud/amazon/rds_snapshot_copy.py [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+
+        if source_region and source_region != region:
+            source_conn = boto.rds.connect_to_region(
+                source_region, **aws_connect_kwargs)
+        else:
+            source_conn = conn
+
+        copy_snapshot(
+            module=module,
+            source_conn=source_conn,
```
Comment (willthames) [inline cloud/amazon/rds_snapshot_copy.py:136]: `source_conn` doesn't seem to be used at all in `copy_snapshot`

File: cloud/amazon/rds_snapshot_copy.py [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+        module.fail_json(msg="Snapshot not available with identifier "
+                             "%s" % (source_db_snapshot_identifier))
+
+    snapshot = conn.copy_dbsnapshot(
+        source_snapshot_id=source_db_snapshot_identifier,
+        target_snapshot_id=target_db_snapshot_identifier
+    )
+
+    module.exit_json(changed=True,
+                     snapshot_id=snapshot.id)
```
Comment (willthames) [inline cloud/amazon/rds_snapshot_copy.py]: this linebreak seems unnecessary

File: cloud/amazon/rds_snapshot_copy.py [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+    required: true
+
+author: "Brock Haywood (@brockhaywood)"
+extends_documentation_fragment:
+    - aws
+    - ec2
+'''
+
+EXAMPLES = '''
+
```
Comment (willthames) [inline cloud/amazon/rds_snapshot_copy.py]: EXAMPLES should not be empty

File: cloud/amazon/rds_snapshot_copy.py [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+                  source_db_snapshot_identifier=None,
+                  target_db_snapshot_identifier=None):
+
+    if not source_db_snapshot_identifier or not target_db_snapshot_identifier:
+        module.fail_json(msg='Source DB Snapshot Id and Target DB Snapshot '
+                             'Id must be provided')
+
+    try:
+        snapshots = conn.get_all_dbsnapshots(
+            snapshot_id=source_db_snapshot_identifier)
```
Comment (willthames) [inline cloud/amazon/rds_snapshot_copy.py]: Ansible doesn't require strict PEP-8 linewidth compliance - not breaking here would improve readability.

**Comment by willthames** [comment]
Please note that amount of feedback doesn't reflect a major concern with this module - it's a good module that just needs a few style/behaviour tweaks.

File: cloud/amazon/rds_snapshot_copy.py (hunk 4 above)
Comment (brockhaywood) [inline cloud/amazon/rds_snapshot_copy.py]: @willthames i based this off the ec2_snapshot module from the core modules repo for consistency. If thats not the convention, will update.

Comment (willthames) [inline cloud/amazon/rds_snapshot_copy.py]: Hmm, that's not how I originally wrote `ec2_snapshot`. Apparently the changes make it "more testable".

Either way, not your fault, and this is not a blocker.

File: cloud/amazon/rds_snapshot_copy.py (hunk 5 above)
Comment (brockhaywood) [inline cloud/amazon/rds_snapshot_copy.py:136]: fixed that bug

**Comment by brockhaywood** [comment]
@willthames i believe i have addressed the feedback provided on this PR

**Comment by willthames** [comment]
The only problem left I see is inconsistent spacing around `=`. It's probably worth following PEP-8 guidelines for spacing if only for consistency sake (I've ignored line length errors (E501) as per Ansible's module development guidelines): 

```
[wthames@wthames amazon (pr-1280)]$ pep8 rds_snapshot_copy.py --ignore=E501
rds_snapshot_copy.py:78:29: E251 unexpected spaces around keyword / parameter equals
rds_snapshot_copy.py:78:31: E251 unexpected spaces around keyword / parameter equals
rds_snapshot_copy.py:103:19: E251 unexpected spaces around keyword / parameter equals
rds_snapshot_copy.py:103:21: E251 unexpected spaces around keyword / parameter equals
rds_snapshot_copy.py:104:26: E251 unexpected spaces around keyword / parameter equals
rds_snapshot_copy.py:104:28: E251 unexpected spaces around keyword / parameter equals
rds_snapshot_copy.py:105:42: E251 unexpected spaces around keyword / parameter equals
rds_snapshot_copy.py:105:44: E251 unexpected spaces around keyword / parameter equals
rds_snapshot_copy.py:106:42: E251 unexpected spaces around keyword / parameter equals
rds_snapshot_copy.py:106:44: E251 unexpected spaces around keyword / parameter equals
rds_snapshot_copy.py:122:29: E251 unexpected spaces around keyword / parameter equals
rds_snapshot_copy.py:122:31: E251 unexpected spaces around keyword / parameter equals
rds_snapshot_copy.py:143:29: E251 unexpected spaces around keyword / parameter equals
rds_snapshot_copy.py:143:31: E251 unexpected spaces around keyword / parameter equals
```

Basically there should be spaces between operators, but not between arguments in a dict.

**Comment by brockhaywood** [comment]
Sheesh, not having my post-commit hooks makes me miss that stuff. Updated

**Comment by willthames** [comment]
passes_guidelines

**Comment by willthames** [comment]
Please squash commits.

**Comment by brockhaywood** [comment]
squashed commits

**Comment by manicminer** [comment]
works_for_me !

A nice-to-have might be a 'wait' parameter, similar to other AWS modules.

**Pull request closed without merging**